### **Detecting Fake News with Python**

Build a Model to accurately Classify news stream as REAL or FAKE

+ Using Scikit-Learn, build a TfidfVectorizer on the Datasets.
+ Initiate a PassiveAggressiveClassifier and fit the Model.
+ Perfomance of Model is measured by Accuracy Matrix and Confusion Matrix 

#### Import Required Libraries

In [70]:
import numpy as np
import pandas as pd
import itertools

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

#### Dataset

The Dataset used for this task is "news.csv"

+ shape is (6335, 4)
+ columns ==> news, title, text, label

Read the Data into Dataframe

In [71]:
df = pd.read_csv("news.csv")
df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [72]:
df.shape

(6335, 4)

#### Get the Labels from the Dataset

In [73]:
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

#### Split the Data into training and testing Set

In [74]:
X_train, X_test, y_train, y_test = train_test_split(df["text"], labels, test_size = 0.2, random_state = 7)

In [75]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(5068,)
(1267,)
(5068,)
(1267,)


#### Initialize TfidfVectorizer

+ Initialize a TfidfVectorizer with stopwords from English language and a maximum document frequency of 0.7
+ Stopwords are the most common words in a language that are to be filtered out before processing the natural language data.

+ A TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features

In [76]:
# initialize a TfidfVectorizer #
tfidf_vectorizer = TfidfVectorizer(stop_words = "english",
                                  max_df = 0.7)

# fit and tranform train set , transform test set #
tfidf_train = tfidf_vectorizer.fit_transform(X_train)
tfidf_test = tfidf_vectorizer.transform(X_test)

In [77]:
tfidf_train

<5068x61651 sparse matrix of type '<class 'numpy.float64'>'
	with 1337098 stored elements in Compressed Sparse Row format>

In [78]:
tfidf_test

<1267x61651 sparse matrix of type '<class 'numpy.float64'>'
	with 322056 stored elements in Compressed Sparse Row format>

In [79]:
tfidf_vectorizer

TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.float64'>, encoding='utf-8',
                input='content', lowercase=True, max_df=0.7, max_features=None,
                min_df=1, ngram_range=(1, 1), norm='l2', preprocessor=None,
                smooth_idf=True, stop_words='english', strip_accents=None,
                sublinear_tf=False, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, use_idf=True, vocabulary=None)

#### PassiveAggressiveClassifier

+ Initialize the PassiveAggressiveClassifier
+ fit this on tfidf_train and y_train
+ Then, predict on the test set from the TfidfVectorizer and Calculate the accuracy with accuracy score

In [80]:
# initialize PassiveAggressiveClassifier #
pac = PassiveAggressiveClassifier(max_iter = 50)
pac.fit(tfidf_train, y_train)

# predict on the test set and calculate accuracy #
y_pred = pac.predict(tfidf_test)
scores = accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(scores*100,2)}%')

Accuracy: 92.58%


The Accuracy of the Model is 92.58%

Print out the confusion matrix to get complete insights about the true and false negatives and positives.

In [81]:
# Build Confusion matrix #

confusion_matrix(y_test, y_pred, labels = ["FAKE", "REAL"])

array([[587,  51],
       [ 43, 586]], dtype=int64)

+ 587 True Positives
+ 586 False Positives
+ 46 False Negatives 
+ 43 False Positives